c++/internal/cc_capnp_library.bzl

```diff
@@ -0,0 +1,150 @@
+
+    gen_dir = ctx.var["GENDIR"]
+    out_dir = gen_dir
+    if src_prefix != "":
+        out_dir = out_dir + "/" + src_prefix
+
+    cc_out = "-o%s:%s" % (ctx.executable._capnpc_cxx.path, out_dir)
+    args = ctx.actions.args()
+    args.add_all(["compile", "--verbose", cc_out])
+    args.add_all(["-I" + inc for inc in includes])
+    args.add_all(["-I", system_include])
+
+    if src_prefix == "":
+        # guess src_prefix for generated files
+        for src in ctx.files.srcs:
+            if src.path.startswith(gen_dir):
+                src_prefix = gen_dir
+                break
+
+    if src_prefix != "":
+        args.add_all(["--src-prefix", src_prefix])
+
+    args.add_all([s for s in ctx.files.srcs])
+
+    # process language extras, starting with files we should include
+    extra_inputs = ctx.files.language_extras or []
+
+    # then compute input flags, taking the parent of the first file in each extra as the include path
+    if len(ctx.attr.language_extras) > 0:
+        for extra in ctx.attr.language_extras:
+            files = extra.files.to_list()
+            if len(files) > 0:
+                args.add_all(["-I", files[0].dirname.removesuffix("/capnp")])
+
+    ctx.actions.run(
+        inputs = inputs + ctx.files._capnpc_cxx + ctx.files._capnpc_capnp + ctx.files._capnp_system + extra_inputs,
+        outputs = ctx.outputs.outs,
+        executable = ctx.executable._capnpc,
+        arguments = [args],
+        mnemonic = "GenCapnp",
+    )
+
+    return [
+        capnp_provider(
+            includes = includes,
+            inputs = inputs,
+            src_prefix = src_prefix,
+        ),
+    ]
+
+_capnp_gen = rule(
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "deps": attr.label_list(providers = [capnp_provider]),
+        "data": attr.label_list(allow_files = True),
+        "outs": attr.output_list(),
+        "src_prefix": attr.string(),
+        "language_extras": attr.label_list(default = []),
+        "_capnpc": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnp_tool"),
+        "_capnpc_cxx": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-c++"),
+        "_capnpc_capnp": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-capnp"),
+        "_capnp_system": attr.label(default = "@capnp-cpp//src/capnp:capnp_system_library"),
+    },
+    output_to_genfiles = True,
+    implementation = _capnp_gen_impl,
+)
+
+_DEFAULT_REPO_NAME = "@capnp-cpp"
+_RUNTIME_REPO_TPL = "%s//src/capnp:capnp_runtime"
+
+
+def cc_capnp_library(
+        name,
+        srcs = [],
+        data = [],
+        deps = [],
+        src_prefix = "",
+        language_extras = [],
+        visibility = None,
+        target_compatible_with = None,
+        repo = _DEFAULT_REPO_NAME,
+        rule = native.cc_library,
```

Comment:
Two very small additions to the macro are these `repo` and `rule` arguments, both with defaults. The `rule` argument will let downstream users customize this macro to use `@rules_cc`, or otherwise wrap the `cc_library` rule defined by `cc_capnp_library`.

The `repo` argument is necessary because Cap'n'Proto now may be installed at a different repo name (than `@capnp-cpp`) within a downstream project's codebase. Using this argument, they can set it to the name they are using so that the runtime dependencies can be resolved properly (`@capnp-cpp//src/capnp_runtime`, etc).